### DISCUSSION TITLE
Cache, Manifests and Result classes

### DISCUSSION THREAD

**Discussion by AlexanderLanin** [ccache/ccache#777] [Development]
I'm trying to grasp how "results" work. It's not trivial...
My current assumptions if you could correct me:

* `Result` can read/write cached results via `CacheEntryReader/Writer` and forward those to a `Reader::Consumer` in case of reading.
* `CacheEntryReader` and `CacheEntryWriter` (although split into two) are basically a specialized version of `File`. They are designed for results and manifests, but could be used for anything that has a header, is compressed and has a checksum.
* `CacheFile`, `CacheEntryReader` and `CacheEntryWriter` basically represent the concept of one CacheEntry. In this specific case it's a file. 
Would it be reasonable to merge these three into `CacheFile` and make it an implementation of a `CacheEntry` interface?
* `Result` works on files, but actually it doesn't care whether that `CacheEntry` is a file. (3 line code change to pass a `CacheEntry` instead of a path).
* `Result` reads in a very specific file format, but offers it on a super generic `Consumer` interface. Feels like some abstraction is missing. Maybe todays `Result` is actually `ResultFile`?
* `CacheEntryReader/Writer` are todays only users of compression (`Compressor`, `Decompressor`, `NullCompressor` etc). However in theory any other implementation of `CacheEntry` could also use these compressors.
* On second thought `CacheEntryReader/Writer` could be abstracted away from files to work on any stream. For example blobs in sqlite. Or blobs via https or whatever. In that case something like `CacheEntryBlob` could be a reasonable name.

What I'm describing here could be seen as prerequisites for #414. In no way is it supposed to replace anything from #414. It's about making what is already there a little more reasonable.

**Comment by jrosdahl** [comment]
> I'm trying to grasp how "results" work. 

:+1:

> * `Result` can read/write cached results via `CacheEntryReader/Writer` and forward those to a `Reader::Consumer` in case of reading.

Yes. `Result` holds the knowledge about the format of a result file.

> * `CacheEntryReader` and `CacheEntryWriter` (although split into two) are basically a specialized version of `File`.

No. They hold the knowledge of the header of cache entries (of which there currently are two types: result and manifest). Just like the code in `Result` and `Manifest` (and compressor and decompressor classes), the fact that they operate on `FILE*` is an implementation detail that will change as part of #414. The current class structure has been designed with that in mind, but it's certainly possible that some more restructuring will be needed or appropriate.

> They are designed for results and manifests, but could be used for anything that has a header, is compressed and has a checksum.

Yes, they can potentially be used for other future cache entry types.

> * `CacheFile`, `CacheEntryReader` and `CacheEntryWriter` basically represent the concept of one CacheEntry. In this specific case it's a file.

Kind of. I commented on `CacheEntryReader` and `CacheEntryWriter` above.

`CacheFile` contains what's needed to know about a file (which notably can be of an unknown type) in the cache for the "visit all files in the (primary file system) cache" use cases (cleanup, recompression, etc). In contrast to `CacheEntry{Reader,Writer}`, `CacheFile` _is_ logically tied to a file and belongs to the "primary file system backend". It would be reasonable to assume that moving that class into the primary file system backend would make sense, while `CacheEntry{Reader,Writer}` or similar code belongs to the part of the application framework that sends or receives data to/from storage backends.

I'm not 100% happy with the name, so suggestions on an alternative name would be welcome.

> Would it be reasonable to merge these three into `CacheFile` and make it an implementation of a `CacheEntry` interface?

I may of course be wrong, but I don't see how that would improve things. The reader and writer states are completely unrelated, so if we would join them into a single data structure, half of the members would be in unused or undefined state. Put differently, an instantiation of a combined `class CacheFile : public CacheEntry` would (depending on use case) either be used for reading or for writing, never both. Contrast this to a typical read/write file abstraction like `FILE` which contains a shared state (current file pointer) for both reading and writing.

> * `Result` works on files, but actually it doesn't care whether that `CacheEntry` is a file. (3 line code change to pass a `CacheEntry` instead of a path).

Already commented above regarding operating on files.

> * `Result` reads in a very specific file format, but offers it on a super generic `Consumer` interface. Feels like some abstraction is missing. Maybe todays `Result` is actually `ResultFile`?

The purpose of `Result::Reader` is specifically to understand the file format and read it into a header and file entries on which consumers operate.

Could you expand on what you think is missing or is strange? Have you found and studied the three consumers?

> * `CacheEntryReader/Writer` are todays only users of compression (`Compressor`, `Decompressor`, `NullCompressor` etc). However in theory any other implementation of `CacheEntry` could also use these compressors.

That's correct.

> * On second thought `CacheEntryReader/Writer` could be abstracted away from files to work on any stream. For example blobs in sqlite. Or blobs via https or whatever.

Yep, that's the idea: #414 will refactor `CacheEntry{Reader,Writer}`, `Result`, `Manifest`, `Compressor`, `Decompressor`, etc., so that they operate on generic streaming I/O interfaces instead of files specifically.

----

Good questions, by the way! I hope this clears things up.

I'm of course open to improvements, I just wanted to convey that at least most parts of the current structure are somewhat thought through and intentional. :slightly_smiling_face:

**Comment by AlexanderLanin** [comment]
How about we start moving files to get them organized a little without changing any of the functionality at the same time? Especially not addressing any issues mentioned in #414. That change would basically just reflect what you summarized above. It would clearly show all usage of these files so it will be clearer what has to be changed in the future.

e.g. (not investigated in detail):
* move CacheFile, cleanup, compress to backend/file
* move CacheEntryReader/Writer, Result, Manifest to backend/generic
* move compression stuff to backend/compression

I don't want to create some future proof structure here! Just sort what is already there so it's easier to reason about it. Directory structure and namespaces will help understand what is going on.

Does that make sense? I can certainly prepare such a branch. I'm not sure it's a good time to do so at the moment due to amount of existing issues and pull requests vs available time?

Further ideas:
* rename CacheFile to FileInfo and extract type() to a free function somewhere else. FileInfo would be path + cached lstat. It's not cache specific in any way.
* Result feels more like a `AllCacheEntriesInOneStream` thingy than a `Result`. Consumer/Writer can be extracted into a separate more generic file.

**Reply by jrosdahl** [reply]
As mentioned in https://github.com/ccache/ccache/pull/753#issuecomment-757537074, I think it's fine to group files where it makes sense but I don't think that we should start moving files around before we have a plan in some simple written form.

To comment on your ideas (and I do understand they're not investigated in detail):

> * move CacheFile, cleanup, compress to backend/file

Could be OK, but it might be a bit strange to move those files to a similar directory before there _is_ a storage backend abstraction.

Allow me to digress slightly. What I meant by "primary file system backend" is something that doesn't exist yet. Or rather, it exists logically but not physically (if you get what I mean) since the code is spread out over several files that also contain code for other things.

Also note the word "primary". My idea for #414 is that there should be two kinds of storage backends: primary and secondary. There can be only one primary storage backend, and it has a richer interface since it needs to include statistics bookkeeping, work with hardlinks/reflinks, etc. There can be zero, one or several secondary storage backends, and a secondary storage backend has a simpler interface (basically only initialize, read, write and deinitialize operations).

My thought has been that the right end to start with is the _secondary_ storage backend interface since (a) it's much easier to define than the primary storage backend interface (and refactoring code into an implementation of it), and (b) it paves the way for storage implementations that people have been wanted for many years. The primary storage backend on the other hand would be nice to have in place for improved code structure, which of course is what you're interested in here, but since it's a much larger task I would personally want to invest more time in secondary backend land simply because I feel it will be a better return of invested time.

> * move CacheEntryReader/Writer, Result, Manifest to backend/generic

Yes, I think it would make sense to put them some kind of storage utility/misc/framework/common place.

> * move compression stuff to backend/compression

Sure. Or to the same place as the files above.

> I'm not sure it's a good time to do so at the moment due to amount of existing issues and pull requests vs available time?

Right, I would like to address a couple of more post-4.0 things before starting this work, but working on plan and taking obvious small steps towards the goal could be started right away.

> Further ideas:
> 
> * rename CacheFile to FileInfo and extract type() to a free function somewhere else. FileInfo would be path + cached lstat. It's not cache specific in any way.

Sounds good. This can be done right away. Would you like to prepare a PR?

> * Result feels more like a `AllCacheEntriesInOneStream` thingy than a `Result`. Consumer/Writer can be extracted into a separate more generic file.

I don't understand your idea here.

`Result` is a namespace which, among other things, contains a reader and a writer of result files, much like `CacheEntry{Reader,Writer}` doesn't represent a stateful cache entry – instead they know how to pack and unpack information. It doesn't represent a stateful result object. The name "result" was chosen since a result file (those with an `R` filename suffix) contains the result of one compilation. To me it sounds appropriate to collect code that knows about the format of such files under the label "result" instead of "all cache entries in one stream". :slightly_smiling_face:

How can Consumer/Writer be made more generic?

**Reply by AlexanderLanin** [reply]
Ah ok, I didn't see that primary / secondary distinction.

My idea is the complete opposite of what you are describing. It was along the lines of:
1) Make current code pretty without really thinking about an API.
2) Repeat 1 until a `FileBackend : public Backend` class has naturally emerged.

That's how I usually refactor code into something that has proper interfaces, collections of classes, etc
